chore: rewrite the Zoe introduction (#311)
Agoric/documentation: main/getting-started/intro-zoe.md

# Introduction to Zoe

::: tip Alpha status
Zoe is currently at Alpha status. It has not yet been
formally tested or hardened. It is not yet of production quality.
:::

## What is Zoe?

> Zoe is Agoric's smart contract framework. You should use Zoe to:

Reviewer: ```suggestion
Zoe is Agoric's smart contract framework. Use Zoe to:
```

* **Run your code on-chain**
* **Mint new digital assets**
* **Credibly trade assets**

> Zoe relies on [ERTP](./ertp-introduction.md), our token standard for fungible and non-fungible
> tokens. 

Reviewer: ```suggestion
Zoe relies on Agoric's [*Electronic Rights Transfer Protocol (ERTP)*](./ertp-introduction.md), our token standard.
```

Author: I think it's not important to specify the full name in the Zoe intro, since it doesn't help explain anything and we also don't use the full name often. ERTP and Electronic Rights Transfer Protocol are not great names, so I'd rather not spend much time on them.

Reviewer: On the one hand, yes. On the other, it's generally considered good writing practice to always expand an acronym on first usage, and there are some people who get irked if a doc doesn't do so (the always exception is if it's either effectively become a word in its own right (ATM for the clear reason, NAACP for less obvious reasons) or if one can reasonably assume everyone knows what it stands for)

Author: Yeah, good point. I suppose there may readers who will feel like they are missing something, even though there is nothing to miss.

> ## Why should I use Zoe?

Reviewer: ```suggestion
## Why use Zoe?
```

### For Users ###

> **Zoe is safer.** Putting digital assets in a smart contract has
> traditionally meant you might lose them. But Zoe guarantees that you
> will either get what you wanted or get a full refund. You will never
> leave a contract empty-handed, even if the smart contract is buggy or
> malicious. 

Reviewer: ```suggestion
**Zoe is safer.** Traditionally, putting digital assets in a smart contract has
risked losing them. But Zoe guarantees you
get either what you wanted or a full refund of the assets you put in. You will never
leave a contract empty-handed, even if the smart contract is buggy or
malicious. 
```

### For Developers ###

> **Zoe is easier.**  Writing a smart contract has traditionally meant
> learning a completely new, untried language. And let's hope you don't
> make any mistakes - if you do, your users might lose millions.

Reviewer: ```suggestion
**Zoe is easier.**  Traditionally, writing a smart contract meant
learning a new, untried language. And don't
make any mistakes - if you do, your users might lose millions.
```

> Zoe smart contracts, on the other hand, are written in a secure subset
> of JavaScript. Moreover, Zoe handles all of the escrowing of digital
> assets provided by your users and the subsequent payout of digital
> assets to your users. **Your contract, even if it's buggy, can't cause
> your users to lose their assets.**

Reviewer: ```suggestion
However, you write Zoe contracts in a secure subset
of JavaScript. Moreover, Zoe automatically escrows all user digital
assets and handles their subsequent payout. **Even a buggy contract can't cause
users to lose their assets.**
```

> ## How do I use Zoe?

Reviewer: ```suggestion
## How to use Zoe
```

There are three ways you can get started using Zoe:

* you can write a new contract and install it on Zoe
* you can create a new `instance` of a contract already installed on
  Zoe
* you can receive an `invitation` to join a contract
  instance that someone else has started.

Agoric has written [a number of example contracts
that you can use](/zoe/guide/contracts/), including:
* a [Uniswap
implementation](/zoe/guide/contracts/multipoolAutoswap.md)
* a [covered
call option contract](/zoe/guide/contracts/covered-call.md)
* an [OTC
Desk market maker
contract](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/otcDesk.js)
* contracts for [minting fungible](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/mintPayments.js) and [non-fungible tokens](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/mintAndSellNFT.js)

> 

Author: Gotcha. I removed the first bit about the three ways. I think we do want people to know about the kind of contracts that can be built on Zoe, with links, early. The reader likely knows what Uniswap is, so that we have something like that, that they would be able to build themselves, is meaningful. And, the links are social proof that we're not talking abstractly and promising more than we can deliver. This is important to a dev audience who has been burned on other hyped projects.

"Example Agoric-written contracts for you to look at and use as templates" isn't right from a product perspective. We want to emphasize composability. That means you'll actually use these contracts - not that you will copy and paste and make your own. Using the term `template` gives the wrong idea. Of course, if anyone writes something new, they will look at other contracts, but we don't need to undermine our composability argument.

The Zoe links are correct. If you start with a `/` it starts from the root, which is `main`.

## Using a Zoe smart contract: an example

Let's imagine that our friend Alice has sent a `Zoe invitation` to
[our wallet](/wallet-api.md). This invitation will let us join Alice's
running contract instance in a particular role. Without an invitation,
we are not able to join.

Compare this to a smart contract on Ethereum. On Ethereum, the smart
contract developer must guard against malicious calls and store an
internal access control list to check whether the message sender is

> allowed to send such a message. Zoe, built on Agoric's [object

Author: Hmm, why? `can send` sounds like whether it is mechanically possible, which gives the wrong impression, since it's more like `may send`. I think `allowed` is correct.

capability](/glossary/#object-capabilities) security model, is simply
easier.

Let's also imagine that this particular invitation is for an
[AtomicSwap
contract](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js),
in which one party puts up some digital assets and sends an invitation
to a second party to complete the exchange. In this case, Alice has
already escrowed her side and is asking us to pay a particular price
to receive her digital assets.

### Inspecting an invitation

So we have an invitation, but how do we use it? We can use Zoe to
inspect the invitation.

> 

Author: sounds good

<<< @/snippets/test-intro-zoe.js#details

::: warning Note

E() is part of the Agoric platform and is used to [call methods on
remote
objects and receive a promise for the result](/distributed-programming.md#communicating-with-remote-objects-using-e). From our perspective, Zoe is a

> remote object, so we must use E().

Author: Yup, good point

> remote object, so we must use E().

Author: I added some context. It's a little funky, but hopefully gets the job done

:::

The invitation contained information about the `installation`, the
installed code object that was used to create the instance. We can use
object identity comparison to quickly check that the installation is
one that we recognize. There is no need to compare line-by-line to

> ensure the code is the same.

Author: This sounds good, thanks!

<<< @/snippets/test-intro-zoe.js#isCorrectCode

However, if we don't recognize the installation, we can inspect the
code directly by calling:

<<< @/snippets/test-intro-zoe.js#inspectCode

Contracts can add their own information to the invitation. In this
case, the Atomic Swap contract adds information about what is being
traded: the `asset`, the amount which Alice has escrowed, and the
`price`, what we must pay to get the asset. Let's imagine that `asset`
was an amount of 3 moola, and `price` was an amount of 7 simoleans.
(Moola and simoleans are made-up currencies for the purpose of this
example.) Amounts are descriptions of digital assets, but have no
value themselves. [For more on amounts, please see the ERTP

> guide.](/ertp/guide/amounts.html#amounts-values-and-brands)

Author: Contracts - it's the code level

Asset and Price.

We said that we are giving 7 simoleans, so we must send 7 simoleans in
the form of a ERTP payment. ([ERTP
payments](/ertp/guide/purses-and-payments.md) are how we transfer
fungible and nonfungible digital assets on the Agoric platform.) Let's assume that we happen to have some simoleans
lying around in a simolean purse. We can withdraw a payment for this
offer, and construct an object using the same Keyword as our `proposal.give`:

<<< @/snippets/test-intro-zoe.js#getPayments

Now let's harden the `proposal` and `payments` objects that we just
created. [Hardening is transitively
freezing](/distributed-programming.md#harden), and for security, we must harden any
objects that get passed to a remote object like Zoe.

<<< @/snippets/test-intro-zoe.js#harden

Now that we have the required pieces, we can make an offer:

<<< @/snippets/test-intro-zoe.js#offer

At this point, Zoe burns our invitation and confirms its validity. Zoe
also escrows all of our payments and represents their value in amounts
as our `current allocation` in the contract.

### Obtaining a UserSeat

> Making an offer as a user gives back a `UserSeat`. This seat
> represents our position in the ongoing contract. We can use this seat
> to do a number of things, including:

Reviewer: ```suggestion
Making an offer as a user returns a `UserSeat`. This seat
represents your position in the ongoing contract instance (your "seat at the table"). 
You can use this seat to:
```

> 1. exit the contract
> 2. get information about our position such as our current allocation
> 3. get our payouts from Zoe.

Reviewer: ```suggestion
1. Exit the contract.
2. Get information about our position such as our current allocation.
3. Get our payouts from Zoe.
```

> Let's check that our offer was successful:

Reviewer: ```suggestion
Check that your offer was successful:
```

<<< @/snippets/test-intro-zoe.js#offerResult

> The AtomicSwap contract gives back this particular message for our
> offer: "The offer has been accepted. Once the contract has been
> completed, please check your payout." Other contracts and offers may
> return something different. The `offerResult` is entirely up to the
> contract.

Reviewer: ```suggestion
In response to your offer, the AtomicSwap contract returns the message: "The offer has been accepted. 
Once the contract has been completed, please check your payout." Other contracts and offers may
return something different. The offer's result is entirely up to the contract.
```

> ### Getting the payouts

Reviewer: ```suggestion
### Getting payouts
```

> Because this was the atomic swap contract, the contract is over once
> the second party escrows the correct assets. Let's get our payout of
> moola using the Keyword we used ('Asset'):

Reviewer: ```suggestion
Because this was an Atomic Swap contract, it is over once
the second party escrows the correct assets to fulfill the offer and do the swap. Get your payout of
moola with the Keyword your used ('Asset'):
```

<<< @/snippets/test-intro-zoe.js#getPayout

> Alice receives her payouts too:

Reviewer: ```suggestion
Alice also receives her payouts:
```

<<< @/snippets/test-intro-zoe.js#alicePayout

## Writing and installing a contract

> We've gone over the process of inspecting an invitation, making an
> offer, receiving a `UserSeat` and getting a payout, but how did Alice
> create the contract instance in the first place?

Reviewer: ```suggestion
Now that you've seen how to participate in a contract instance, let's look at how
you'd create a contract and its instances.
```

> Let's pretend that Alice wrote the contract from scratch, even
> AtomicSwap is already written as one of our example contracts ([see
> the full AtomicSwap code
> here](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js)).
> Note: All Zoe contracts must have a certain format, shown below.

Reviewer: ```suggestion
Let's pretend Alice wrote that contract from scratch, even though
AtomicSwap is one of Agoric's example contracts ([see
the full AtomicSwap code
here](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js)).
Note: All Zoe contracts must have this format:
```

::: details Show contract format
<<< @/snippets/contract-format.js#contractFormat
:::

> 
> 
> Let's imagine that Alice fills in this code with the particulars of
> the [AtomicSwap implementation](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js). 
> 
> To install this particular code, Alice must first bundle it off-chain, meaning
> that the code and its imports are flattened together:

Reviewer: ```suggestion
Alice fills in this code template with [AtomicSwap's particulars](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js). 
To install this particular code, Alice first must bundle it off-chain, meaning the code and its imports are flattened together:
```

> <<< @/snippets/test-intro-zoe.js#importBundleSource

Reviewer 2: This is empty. Was something supposed to go here?

Author: Oh wow, I think `yarn lint-fix` really messed it up. I checked this before linting, but not after. Thanks, I'll fix.

Author: Fixed

<<< @/snippets/test-intro-zoe.js#bundle

Then Alice must install it on Zoe:

<<< @/snippets/test-intro-zoe.js#install

> The return value is an `installation`, which we saw earlier. This is
> an object that identifies a particular piece of code installed on Zoe.
> It can be compared to other installations, and we can call
> `E(atomicSwapInstallation).getBundle()` to read the code itself.

Reviewer: ```suggestion
The return value is an `installation`, which we saw earlier. It is
an object identifying a particular piece of code installed on Zoe.
It can be compared to other installations, and you can call
`E(atomicSwapInstallation).getBundle()` to see the code itself.
```

### Creating an instance

> Now Alice can use the installation to create a new instance. She must
> also tell Zoe about the ERTP issuers that she wants to use, by
> specifying their role with Keywords. Alice was escrowing moola, so she
> will use the keyword `Asset` to label the `moolaIssuer`. She wanted
> simoleans, so she will use the keyword `Price` to label the `simoleanIssuer`. This particular contract ([see permalink to line
> 58](https://github.com/Agoric/agoric-sdk/blob/23c3f9df56940230e21a16b4861f40197192fdea/packages/zoe/src/contracts/atomicSwap.js#L58))
> returns a `creatorInvitation` as the result of `startInstance`, so that's what Alice can use to make
> her offer.

Reviewer: ```suggestion
Now Alice uses the installation to create a new instance. She must
also tell Zoe about the ERTP issuers she wants to use, by
specifying their role with Keywords. Alice was escrowing moola, so she uses the keyword `Asset` to label the `moolaIssuer`. She wanted
simoleans, so she uses the keyword `Price` to label the `simoleanIssuer`. When you create a new instance of the contract ([see permalink to line
58](https://github.com/Agoric/agoric-sdk/blob/23c3f9df56940230e21a16b4861f40197192fdea/packages/zoe/src/contracts/atomicSwap.js#L58))
by calling `startInstance()`, it returns a `creatorInvitation`. Alice uses this to make her offer; even the instance's creator needs to have an invitation to it to participate in it.
```

<<< @/snippets/test-intro-zoe.js#startInstance

> As per the [Atomic Swap contract

Reviewer: I think this is a combination of redundant/conflicting with the previous paragraph? Rephrase or combine the two?

Author: Nope, this is not covered before, and it's not conflicting. This is talking about the reader's invitation, not Alice's invitation.

code](https://github.com/Agoric/agoric-sdk/blob/23c3f9df56940230e21a16b4861f40197192fdea/packages/zoe/src/contracts/atomicSwap.js#L50),

> Alice gets an invitation as a result of her offer. It was this
> invitation that she sends to the counter-party.

Reviewer 2: inconsistent tense
```suggestion
Alice gets an invitation as a result of her offer. This
is the invitation she sends to the counter-party.
```

Author: Good catch.

<<< @/snippets/test-intro-zoe.js#aliceOffer

> ## Two sides of Zoe: Zoe Service and ZCF (Zoe Contract Facet)

Reviewer: ```suggestion
## Zoe's two sides: Zoe Service and Zoe Contract Facet (ZCF)
```

You may have noticed that the contract code's `start` method had a
parameter called `zcf`. This is the Zoe Contract Facet. Zoe actually
has two sides: the Zoe Service, which we've seen the users interact with, and
the Zoe Contract Facet (ZCF), which is accessible to the contract

> code. Note that users do not have access to ZCF, but contract code
> does have access to the Zoe Service.

Reviewer 2: ```suggestion
code. Note that users do not have access to ZCF, but contract code
can get access to the Zoe Service.
```

> You may have noticed that the contract code's `start` method had a
> parameter called `zcf`. This is the Zoe Contract Facet. Zoe actually
> has two sides: the Zoe Service, which we've seen the users interact with, and
> the Zoe Contract Facet (ZCF), which is accessible to the contract
> code. Note that users do not have access to ZCF, but contract code
> can get access to the Zoe Service.

Reviewer: ```suggestion
You may have noticed the contract code's `start` method had a `zcf`
parameter . This is the Zoe Contract Facet. Zoe
has two sides: the Zoe Service, which you've seen users interact with, and
the Zoe Contract Facet (ZCF), which is accessible to the contract
code. Note that users have access to the Zoe service, but do not have access to the ZCF.
While contract code has access to the ZCF *and* can get access to the Zoe Service. 

The overall Zoe API also includes some very useful Zoe Helper methods, which both you and 
contract code can use.
```

> To learn more about the Zoe Service and Zoe Contract Facet APIs, [see

Reviewer: ```suggestion
To learn more about the Zoe Service, Zoe Contract Facet, and Zoe Helper APIs, [see
```

our Zoe API documentation](/zoe/api/).

## Next steps

If you want to dive deeper into how Zoe works and what you can do, go
to the [Zoe Guide](/zoe/guide/README.md). 

> To learn more about the AtomicSwap contract in particular, you can
> [read its documentation](/zoe/guide/contracts/atomic-swap.md) and take [a look at the source code](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js).

Reviewer: ```suggestion
To learn more about the AtomicSwap contract, you can
[read its documentation](/zoe/guide/contracts/atomic-swap.md) and [look at its source code](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contracts/atomicSwap.js).
There are several other example contracts for different transaction types in 
the [Contracts folder](/zoe/guide/contracts/)
```

To start building Zoe contracts and applications (dapps), follow the
instructions in [Starting a
Project](/getting-started/start-a-project.md) after installing the

> prerequisites.

Reviewer: ```suggestion
Project](/getting-started/start-a-project.md) after [installing the
prerequisites](/getting-started/before-using-agoric.md).

> To explore the Zoe Service and Zoe Contract Facet APIs, see the [Zoe
> API documentation here](/zoe/api/).

Reviewer: ```suggestion
To explore the Zoe Service, Zoe Contract Facet, and Zoe Helper APIs, see the [Zoe
API documentation](/zoe/api/).
```

::: tip Alpha status
Zoe is currently at Alpha status. It has not yet been
formally tested or hardened. It is not yet of production quality.
:::

## What is Zoe?

Zoe is Agoric's smart contract framework. You should use Zoe to:

* **Run your code on-chain**
* **Mint new digital assets**
* **Credibly trade assets**

Zoe relies on [ERTP](./ertp-introduction.md), our token standard for fungible and non-fungible
tokens. 

## Why should I use Zoe?

### For Users ###

**Zoe is safer.** Putting digital assets in a smart contract has
traditionally meant you might lose them. But Zoe guarantees that you
will either get what you wanted or get a full refund. You will never
leave a contract empty-handed, even if the smart contract is buggy or
malicious. 

### For Developers ###

**Zoe is easier.**  Writing a smart contract has traditionally meant
learning a completely new, untried language. And let's hope you don't
make any mistakes - if you do, your users might lose millions.

Zoe smart contracts, on the other hand, are written in a secure subset
of JavaScript. Moreover, Zoe handles all of the escrowing of digital
assets provided by your users and the subsequent payout of digital

> assets to your users. **Your contract, even if it's buggy, can't cause
> your users to lose their assets.**

Reviewer 2: Just an idea.
```suggestion
assets to your users. **Your contract, even if it has bugs due to a library you 
relied on, can't cause your users to lose their assets.**
```

Author: I like this concept, but I think it might be too wordy for this particular sentence. Is there a good way to still include?

Reviewer 2: This is shorter. Does it work here?

"even if you relied on a buggy contract"

Reviewer 2: "even if you relied on a buggy library"

Author: I think this is changing the meaning significantly. I'm trying to say, 'even if the contract *you wrote* is buggy, your users are fine" I think bringing up other contracts and libraries is a good point, but it doesn't seem to fit well.

Reviewer 2: My goal was to seem less accusatory. (Some) People don't immediately accept that they need to protect customers from their own mistakes, but I think the argument that code you rely on may have unnoticed bugs or bugs introduced later is becoming more accepted.

Author: Hmm, but I think some people aren't relying on any libraries or other code, so that argument doesn't work for them. But I do think that most responsible smart contract devs would be worried about a Parity-like problem.

example.) Amounts are descriptions of digital assets, but have no
value themselves. [For more on amounts, please see the ERTP
guide.](/ertp/guide/amounts.html#amounts-values-and-brands)

### Making an offer

Now that we've checked out the invitation, we can make an offer.

There are three required parts to an offer:
* a Zoe invitation
* a proposal
* ERTP payments of digital assets, to be escrowed

The `proposal` is a statement of what you want from the offer, and
what you are giving in return. Zoe will use the proposal as an
invariant to ensure that we don't lose our assets in the trade. This
invariant is known as offer safety.

We can use the `asset` and `price` amounts from the invitation to make
our proposal: 

```js
const proposal = {
  want: { Asset: asset }, // asset: 3 moola
  give: { Price: price }, // price: 7 simoleans
};
```

Proposals must use Keywords, which are keys that are ASCII and
capitalized. In this case, the keywords to use [were determined by the
contract
code](https://github.com/Agoric/agoric-sdk/blob/23c3f9df56940230e21a16b4861f40197192fdea/packages/zoe/src/contracts/atomicSwap.js#L29):
Asset and Price.

> We said that we are giving 7 simoleans, so we must send 7 simoleans in
> the form of a ERTP payment. ([ERTP

Reviewer 2: ```suggestion
We said that we are giving 7 simoleans, so zoe will ensure that we include 7 simoleans in the 
the form of a ERTP payment. ([ERTP
```

Author: Hmm, I think this sounds like Zoe is helping us make the payment, rather than something that we must do and which Zoe will enforce. I didn't mean to say anything about Zoe here, but more that we need to make a payment.
